lib/lints/avoid_unnecessary_setstate/visitor/avoid_unnecessary_setstate_visitor.dart

```diff
@@ -0,0 +1,154 @@
+  Iterable<MethodInvocation> get setStateInvocations => _setStateInvocations;
+
+  /// All class methods invocations
+  Iterable<MethodInvocation> get classMethodsInvocations =>
+      _classMethodsInvocations;
+
+  @override
+  void visitClassDeclaration(ClassDeclaration node) {
+    super.visitClassDeclaration(node);
+
+    final type = node.extendsClause?.superclass.type;
+    if (type == null || !isWidgetStateOrSubclass(type)) {
+      return;
+    }
+
+    final declarations = node.members.whereType<MethodDeclaration>().toList();
+    final classMethodsNames =
+        declarations.map((declaration) => declaration.name.lexeme).toSet();
+    final bodies = declarations.map((declaration) => declaration.body).toList();
+    final methods = declarations
+        .where((member) => _checkedMethods.contains(member.name.lexeme))
+        .toList();
+    final restMethods = declarations
+        .where((member) => !_checkedMethods.contains(member.name.lexeme))
+        .toList();
+
+    final visitedRestMethods = <String, bool>{};
+
+    for (final method in methods) {
+      final visitor = _MethodVisitor(classMethodsNames, bodies);
+      method.visitChildren(visitor);
+
+      _setStateInvocations.addAll(visitor.setStateInvocations);
+      _classMethodsInvocations.addAll(
+        visitor.classMethodsInvocations
+            .where(
+              (invocation) => _containsSetState(
+                visitedRestMethods,
+                classMethodsNames,
+                bodies,
+                restMethods.firstWhere(
+                  (method) => method.name.lexeme == invocation.methodName.name,
+                ),
+              ),
+            )
+            .toList(),
+      );
+    }
+  }
+
+  bool _containsSetState(
+    Map<String, bool> visitedRestMethods,
+    Set<String> classMethodsNames,
+    Iterable<FunctionBody> bodies,
+    MethodDeclaration declaration,
+  ) {
+    final type = declaration.returnType?.type;
+    if (type != null && (type.isDartAsyncFuture || type.isDartAsyncFutureOr)) {
+      return false;
+    }
+
+    final name = declaration.name.lexeme;
+    if (visitedRestMethods.containsKey(name) && visitedRestMethods[name]!) {
+      return true;
+    }
+
+    final visitor = _MethodVisitor(classMethodsNames, bodies);
+    declaration.visitChildren(visitor);
+
+    final hasSetState = visitor.setStateInvocations.isNotEmpty;
+
+    visitedRestMethods[name] = hasSetState;
+
+    return hasSetState;
+  }
+}
+
+/// AST Visitor which finds all setState invocations and checks if they are
+/// necessary
+class _MethodVisitor extends RecursiveAstVisitor<void> {
+  final Set<String> classMethodsNames;
+  final Iterable<FunctionBody> bodies;
+
+  final _setStateInvocations = <MethodInvocation>[];
+  final _classMethodsInvocations = <MethodInvocation>[];
+
+  Iterable<MethodInvocation> get setStateInvocations => _setStateInvocations;
+  Iterable<MethodInvocation> get classMethodsInvocations =>
+      _classMethodsInvocations;
```

Comment:
Please, apply the correct formatting.

Reply:
This line is automatically formatted by Android Studio according to the maxLineLength - 80 rule.
Is it correct?

Reply:
I mean we should have an empty line between getters. Android Studio should automatically apply this formatting.